ddtrace/contrib/pytest/_atr_utils.py

```diff
@@ -0,0 +1,274 @@
+from ddtrace.internal.test_visibility.api import InternalTest
+
+
+log = get_logger(__name__)
+
+
+class _ATR_RETRY_OUTCOMES:
+    ATR_ATTEMPT_PASSED = "dd_atr_attempt_passed"
+    ATR_ATTEMPT_FAILED = "dd_atr_attempt_failed"
+    ATR_ATTEMPT_SKIPPED = "dd_atr_attempt_skipped"
+    ATR_FINAL_PASSED = "dd_atr_final_passed"
+    ATR_FINAL_FAILED = "dd_atr_final_failed"
+
+
+_FINAL_OUTCOMES: t.Dict[TestStatus, str] = {
+    TestStatus.PASS: _ATR_RETRY_OUTCOMES.ATR_FINAL_PASSED,
+    TestStatus.FAIL: _ATR_RETRY_OUTCOMES.ATR_FINAL_FAILED,
+}
+
+
+def atr_handle_retries(
+    test_id: InternalTestId,
+    item: pytest.Item,
+    when: str,
+    original_result: pytest_TestReport,
+    test_outcome: _TestOutcome,
+):
+    # Overwrite the original result to avoid double-counting when displaying totals in final summary
+    if when == "call":
+        if test_outcome.status == TestStatus.FAIL:
+            original_result.outcome = _ATR_RETRY_OUTCOMES.ATR_ATTEMPT_FAILED
+        return
+    if InternalTest.get_tag(test_id, "_dd.ci.atr_setup_failed"):
+        log.debug("Test item %s failed during setup, will not be retried for Early Flake Detection")
+        return
+    if InternalTest.get_tag(test_id, "_dd.ci.atr_teardown_failed"):
+        # NOTE: tests that passed their call but failed during teardown are not retried
+        log.debug("Test item %s failed during teardown, will not be retried for Early Flake Detection")
+        return
+
+    atr_outcome = _atr_do_retries(item)
+
+    final_report = pytest_TestReport(
+        nodeid=item.nodeid,
+        location=item.location,
+        keywords=item.keywords,
+        when="call",
+        longrepr=None,
+        outcome=_FINAL_OUTCOMES[atr_outcome],
+    )
+    item.ihook.pytest_runtest_logreport(report=final_report)
+
+
+def atr_get_failed_reports(terminalreporter: _pytest.terminal.TerminalReporter) -> t.List[pytest_TestReport]:
+    return terminalreporter.getreports(_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_FAILED)
+
+
+def _atr_do_retries(item: pytest.Item) -> TestStatus:
+    outcomes = RetryOutcomes(
+        PASSED=_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_PASSED,
+        FAILED=_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_FAILED,
+        SKIPPED=_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_SKIPPED,
+        XFAIL=_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_PASSED,
+        XPASS=_ATR_RETRY_OUTCOMES.ATR_ATTEMPT_FAILED,
+    )
+
+    test_id = _get_test_id_from_item(item)
+
+    while InternalTest.atr_should_retry(test_id):
+        retry_num = InternalTest.atr_add_retry(test_id, start_immediately=True)
+
+        with set_retry_num(item.nodeid, retry_num):
+            retry_outcome = _get_outcome_from_retry(item, outcomes)
+
+        InternalTest.atr_finish_retry(
+            test_id, retry_num, retry_outcome.status, retry_outcome.skip_reason, retry_outcome.exc_info
+        )
+
+    return InternalTest.atr_get_final_status(test_id)
+
+
+def _atr_write_report_for_status(
```

Comment:
Isn't this function (and the next one) literally identical to the one from EFD?

Reply:
It is indeed. I think there are a few more opportunities for DRY-oriented refactors. I don't know if it's that important to do right now.